src/ksym.rs

```diff
@@ -91,29 +91,69 @@ impl KSymResolver {
         }
     }
 
-    pub fn find_address_ksym(&self, addr: u64) -> Option<&Ksym> {
+    pub fn find_address_ksym(&self, addr: u64) -> Vec<&Ksym> {
         let mut l = 0;
+        let mut dropped_l = 0;
         let mut r = self.syms.len();
 
         if !self.syms.is_empty() && self.syms[0].addr > addr {
-            return None;
+            return vec![];
         }
 
         while l < (r - 1) {
             let v = (l + r) / 2;
             let sym = &self.syms[v];
 
-            if sym.addr == addr {
-                return Some(sym);
-            }
             if addr < sym.addr {
                 r = v;
             } else {
+                if self.syms[l].addr != sym.addr {
+                    dropped_l = l;
+                }
                 l = v;
             }
         }
+        assert!(
+            l == 0 || self.syms[l].addr <= addr,
+            "left addr <= addr < right addr"
+        );
+        assert!(
+            r == self.syms.len() || addr < self.syms[r].addr,
+            "left addr <= addr < right addr"
+        );
+
+        // Seek `l` to the first symbol with the same addr as the
+        // symbol at the current position given by l.
+        let start_addr = self.syms[l].addr;
+        if self.syms[dropped_l].addr == start_addr {
+            assert_eq!(dropped_l, 0);
+            l = dropped_l;
+        } else {
+            // Do binary searching in the range (dropped_l, l] to find the
+            // first symbol with the same address as lth symbol.
+            while dropped_l < (l - 1) {
```

Comment:
so this function can be logically split into two parts:
1) find rightmost ksym with ksym.addr <= addr
2) starting from that symbol, append all symbols that have ksym.addr <= addr and stop at the first ksym.addr > addr

Is that right? This should be doable with one simple and elegant binary search, something like this (in C and sort of pseudo-code, but hopefully will give an idea, but note that I haven't even tried to compile this):

```
int n = self.syms.len();
int l = 0, r = n - 1;
while (l < r) {
  int m = l + (r - l + 1) / 2;
  if (self.syms[m].addr <= addr)
    l = m;
  else
    r = m - 1;
}
for (; l < n && self.syms[l].addr <= addr; l++) {
    results.push(&self.syms[l]);
}
```

Note that you don't need any extra pre- or post-checks, this should handle all the cases cleanly.

Few points here that are missing from current implementation:

a) the invariant is that index `l` always points to ksym with ksym.addr <= addr, and binary search makes sure we always try to pull this left index to the right, if possible. One corner case is when leftmost ksym already has too big addr. We can specially check that condition before the loop, but in this case it is not necessary because the loop that add elements tests this condition anyway. So binary search will end with l == r == 0 and following loop will add no elements.
b) `while (l < r)` makes sure we don't miss any elements, we'll stop only when l == r, you code has `l < r - 1` which while might be correct if you take into account all the other checks, doesn't inspire confidence at all
c) `int m = l + (r - l + 1) / 2;` makes sure that if we have l == r-1, we always try right element and if it satisfies condition we update to point to that right element
d)  `r = m - 1;` in combination with above means we always make progress

There are literally 5 meaningful lines in any binary search implementation and each line is extremely important. But also slight variations of those 5 lines define different variations of what you need to find (leftmost vs rightmost element with some property, etc)

Reply:
After the while-loop, all symbols with an address that is less or equal to `addr` will be placed before `l` or at `l`, and all address bigger than `addr` will at the right side of `l`.  The address of the corresponding symbols of `addr` is always less or equal to `addr`. That means the corresponding symbols will be at the left side of `l`. `l` will point to the "last" one of the symbols with the same address.

Does that make sense?

Reply:
After the while-loop, all symbols with addresses that are less than or equal to `addr` will be after the left size of `l` except the last one.  All symbols with addresses that are bigger than `addr` will be at the right size of `l`.  And, the address of the corresponding symbols of `l` is always less or equal to `addr`.  That mean `l` points to the last one, the leftmost, of the corresponding symbols.

Does that make sense?

We can push symbols into the vector in the reversed order instead.  Then, we don't need to seek `l` to the first matched symbol.

Reply:
Not exactly clear which part of my comment you are replying to.

But looking at the original code again, I think I got the problem slightly wrong assuming we want to find all the symbols that might cover provided addr. Now I think we only want to find exact matches, right? So it requires only a slight change to my example:

```
int n = self.syms.len();
int l = 0, r = n - 1;
while (l < r) {
  int m = l + (r - l) / 2;
  if (self.syms[m].addr < addr)
    l = m + 1;
  else
    r = m;
}
for (; l < n && self.syms[l].addr == addr; l++) {
    results.push(&self.syms[l]);
}
```

I updated few details:
a) `int m = l + (r - l) / 2;` so we try left element once we get to two-element range
b)
```
  if (self.syms[m].addr < addr)
    l = m + 1;
  else
    r = m;
```
Here invariant is that `r` contains leftmost symbol with `ksym.addr >= addr`

c) `for (; l < n && self.syms[l].addr == addr; l++) {` so that we only add ksyms with exact addr match, if there is any.

As I said, there are few parts to each binary search implementation which define exact algorithm details, but the overall structure is always the same.

Please let me know if you think this solves the problem.

My big point here is that nested binary searches are not good and elegant solution, so we should avoid it if we can.

Reply:
kallsyms contains (start address, symbol name) pairs. The function is to find a symbol A followed by symbol B where the address of A <= target_address < the address of B, and includes all symbols with the same start address as A.  So, we are not doing exactly matching.

Reply:
The optimization here seems to be very confusing to people.  Should we go back to the most simple way by doing a backward linear search?

Reply:
> kallsyms contains (start address, symbol name) pairs, and we sort them. The function is to find a symbol A followed by symbol B where the address of A <= target_address < the address of B, and includes all symbols with the same start address as A. So, we are not doing exactly matching.

ok, then my first example should do the job and replace this entire function. The only extra thing you'd need is in the for() loop that goes after while() loop, in addition to checking that ksym.addr <= addr, we need to check that next ksym.addr starts after addr (because kallsyms doesn't store the size of the symbol).

WDYT?

Reply:
@ThinkerYzu1 ok, after re-reading your last reply and the one quoted below I now understand what you are trying to say. If I actually understand :) then we need to change for() loop after while() loop in my first example to go backwards as long as `ksym[l].addr <= addr && addr < ksym[l + 1].addr`. Is that right? So what you said, just add elements in backwards order.

> After the while-loop, all symbols with addresses that are less than or equal to `addr` will be at the left side of `l` except the last one. All symbols with addresses that are bigger than `addr` will be at the right side of `l`. And, the address of the corresponding symbols of `l` is always less or equal to `addr`. That mean `l` points to the last one, the leftmost, of the corresponding symbols.
> 
> Does that make sense?
> 
> We can push symbols into the vector in the reversed order instead. Then, we don't need to seek `l` to the first matched symbol.

yep.

My **main point** I was trying to make in this comment thread is that this is the problem that should be solvable with one clean binary search loop. Nested loop and all the extra checks should be avoided if possible.

WDYT?

Reply:
Agree! Let's add symbols in backward order after the first loop.

Reply:
fixed